### PR TITLE
initial backend deployment

### PR DESCRIPTION
**Pull request by umani-wg**
SAM template and cfn templates for backend deployment.

### CODE REVIEW

File: cfn/iam_policy.yaml [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+            AssumeRolePolicyDocument:
+                Version: "2012-10-17"
+                Statement:
+                    - Effect: Allow
+                      Principal:
+                          Service:
+                              - lambda.amazonaws.com
+                      Action:
+                          - sts:AssumeRole
+            RoleName: !Sub "${RoleName}-${Stage}"
```
Comment (vinsharma-wg) [inline cfn/iam_policy.yaml:28]: Can we use Stage as prefix?

File: cfn/iam_policy.yaml [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+Description: CloudFormation Template for an AWS IAM Role
+
+Parameters:
+    RoleName:
+        Type: String
+        Description: "Name of the IAM role"
+        MinLength: 1
+        MaxLength: 80
+        Default: "lambda-iam-role"
+    Stage:
```
Comment (vinsharma-wg) [inline cfn/iam_policy.yaml:11]: Instead of using the name please use environment. This is the convention we follow

File: cfn/dynamodb.yaml [hunk 3]
```diff
@@ -0,0 +1,41 @@
+AWSTemplateFormatVersion: "2010-09-09"
+Description: >
+    This CloudFormation template creates a DynamoDB table.
+Parameters:
+    TableName:
+        Type: String
+        Description: The name of the DynamoDB table
+    Stage:
```
Comment (vinsharma-wg) [inline cfn/dynamodb.yaml:8]: Instead of using the name please use environment. This is the convention we follow

File: .gitignore [hunk 4]
```diff
@@ -0,0 +1,2 @@
+config.py
+src/common/python/lib
```
Comment (vinsharma-wg) [inline .gitignore:2]: why is this path here. Please add __pycache__ in gitignore also

File: cfn/cognito.yaml [hunk 5]
```diff
@@ -0,0 +1,61 @@
+AWSTemplateFormatVersion: "2010-09-09"
+Description: CloudFormation Template for an AWS Cognito User Pool
+Parameters:
+    Stage:
+        Type: String
```
Comment (vinsharma-wg) [inline cfn/cognito.yaml:5]: Instead of using the name please use environment. This is the convention we follow

File: cfn/dynamodb.yaml [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        Type: String
+        Description: The name of the stage (e.g., development, production)
+
+Resources:
+    DynamoDBTable:
+        Type: AWS::DynamoDB::Table
+        Properties:
+            TableName: !Sub "${TableName}-${Stage}"
```
Comment (vinsharma-wg) [inline cfn/dynamodb.yaml:16]: Can we use Stage as prefix?

File: cfn/iam_policy.yaml [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+            Policies:
+                - PolicyName: CloudWatchLogsPolicy
+                  PolicyDocument:
+                      Version: "2012-10-17"
+                      Statement:
+                          - Effect: Allow
+                            Action:
+                                - logs:CreateLogStream
+                                - logs:PutLogEvents
+                            Resource: "*"
```
Comment (vinsharma-wg) [inline cfn/iam_policy.yaml:38]: Can we avoid "*" ?

File: cfn/static_content_bucket.yaml [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+            Bucket: !Ref VMSStaticContentBucket
+            PolicyDocument:
+                Version: "2012-10-17"
+                Statement:
+                    - Effect: Allow
+                      Principal:
+                          AWS: !Ref "AWS::AccountId"
+                      Action:
+                          - "s3:PutObject"
+                          - "s3:GetObject"
```
Comment (vinsharma-wg) [inline cfn/static_content_bucket.yaml:35]: When we have to get the object. We need more permissions please check that

File: cfn/static_content_bucket.yaml [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+                      Resource:
+                          !Join ["", ["arn:aws:s3:::", !Ref BucketName, "/*"]]
+                    - Effect: Allow
+                      Principal:
+                          Service:
+                              - "s3.amazonaws.com"
+                      Action:
+                          - "s3:ListBucket"
+                      Resource: !Join ["", ["arn:aws:s3:::", !Ref BucketName]]
```
Comment (vinsharma-wg) [inline cfn/static_content_bucket.yaml:44]: Can we move the related permissions to common block

File: run.py [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+        f"--stack-name {env_variables.get('stack_name')} "
+        "--capabilities CAPABILITY_IAM CAPABILITY_NAMED_IAM "
+        f"--parameter-overrides Stage={env_variables.get('stage_name')}"
+    )
+    logger.info("Deploying SAM application...")
+    run_command(deploy_command)
+
+
+def main():
+    deploy_stack(**cognito_stack, action="update")
```
Comment (vinsharma-wg) [inline run.py:119]: Can this deployment be successful at first time?

File: cfn/static_content_bucket.yaml (hunk 9 above)
Comment (vinsharma-wg) [inline cfn/static_content_bucket.yaml:44]: Don't use import and export. When you create a bucket you will get a stack id use a Cloud Formation function to get the output using the stack name or id.

File: src/common/python/vms_layer/db_helper.py [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+ 
+    def delete_item(self, key):
+        """Delete an item from the table"""
+        response = self.table.delete_item(Key=key)
+        return response
+ 
+    def query_items(self, key_condition_expression, expression_attribute_values, page_size=10, starting_token=None):
+        """Query items from the table"""
+        if starting_token:
+            response = self.table.query(
```
Comment (vinsharma-wg) [inline src/common/python/vms_layer/db_helper.py:39]: There is no exception handling in this file

Comment (vinsharma-wg) [reply src/common/python/vms_layer/db_helper.py:39]: Add batch get and batch add based on the requirements.

File: src/common/python/vms_layer/utils/date_helper.py [hunk 12]
```diff
@@ -0,0 +1,29 @@
+"""Helper Class for date conversions."""
+import time
+
+
+def date_To_epoch(date):
+    """Converts a datetime string to epoch time."""
+    return int(time.mktime(time.strptime(date, "%Y-%m-%dT%H:%M:%S")))
+
+
+def epoch_To_date(epoch):
```
Comment (vinsharma-wg) [inline src/common/python/vms_layer/utils/date_helper.py:10]: why is To capitalized?

File: src/common/python/vms_layer/utils/date_helper.py [hunk 13]
```diff
@@ -0,0 +1,29 @@
+"""Helper Class for date conversions."""
+import time
+
+
+def date_To_epoch(date):
```
Comment (vinsharma-wg) [inline src/common/python/vms_layer/utils/date_helper.py:5]: why is To capitalized?

File: cfn/iam_policy.yaml [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+                - PolicyName: S3Policy
+                  PolicyDocument:
+                      Version: "2012-10-17"
+                      Statement:
+                          - Effect: Allow
+                            Action:
+                                - s3:GetObject
+                                - s3:PutObject
+                            Resource: !Join ["", [!ImportValue BucketArn, "/*"]]
```
Comment (vinsharma-wg) [inline cfn/iam_policy.yaml:47]: Please don't use import export. Instead pass the BuckertArn in the parameters and refer it here

File: run.py [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+    "stage_name": config.STAGE_NAME,
+    "stack_name": "api-gateway-lambda-sam",
+    "s3_bucket": config.BUCKET_NAME,
+}
+cognito_stack = {
+    "stack_name": "cognito-stack",
+    "template_body_url": "cfn/cognito.yaml",
+    "parameters": [
+        {"ParameterKey": "Stage", "ParameterValue": env_variables.get("stage_name")},
+        {"ParameterKey": "UserPoolName", "ParameterValue": "vms-user-pool"},
```
Comment (vinsharma-wg) [inline run.py:20]: instead of using vms can we use visitor-mgmt-sys?

File: run.py [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+        },
+    ],
+    "capabilities": ["CAPABILITY_IAM"],
+}
+dynamodb_stack = {
+    "stack_name": "dynamodb-stack",
+    "template_body_url": "cfn/dynamodb.yaml",
+    "parameters": [
+        {"ParameterKey": "Stage", "ParameterValue": env_variables.get("stage_name")},
+        {"ParameterKey": "TableName", "ParameterValue": "vms-database"},
```
Comment (vinsharma-wg) [inline run.py:33]: instead of using vms-database can we use visitor-mgmt-sys-db?

File: run.py [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+    ],
+    "capabilities": ["CAPABILITY_IAM"],
+}
+iam_stack = {
+    "stack_name": "iam-stack",
+    "template_body_url": "cfn/iam_policy.yaml",
+    "parameters": [
+        {"ParameterKey": "Stage", "ParameterValue": env_variables.get("stage_name")},
+        {"ParameterKey": "RoleName", "ParameterValue": "vms-lambda-role-common"},
```
Comment (vinsharma-wg) [inline run.py:42]: instead of using vms can we use visitor-mgmt-sys?

File: run.py [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+    ],
+    "capabilities": ["CAPABILITY_IAM", "CAPABILITY_NAMED_IAM"],
+}
+
+static_content_bucket_stack = {
+    "stack_name": "static-content-bucket-stack",
+    "template_body_url": "cfn/static_content_bucket.yaml",
+    "parameters": [
+        {"ParameterKey": "BucketName", "ParameterValue": "vms-static-content-test"}
```
Comment (vinsharma-wg) [inline run.py:51]: instead of using vms can we use visitor-mgmt-sys?

File: cfn/iam_policy.yaml (hunk 14 above)
Comment (vinsharma-wg) [reply cfn/iam_policy.yaml:47]: I think we need permissions for dynamoDB as well

File: samconfig.toml [hunk 19]
```diff
@@ -0,0 +1,10 @@
+version = 0.1
+[default.deploy.parameters]
+stack_name = "test-api-gateway"
+resolve_s3 = true
+s3_prefix = "test-api-gateway"
+region = "ap-south-1"
+confirm_changeset = true
+capabilities = "CAPABILITY_IAM"
+parameter_overrides = "Stage=\"dev\""
+image_repositories = []
```
Comment (vinsharma-wg) [inline samconfig.toml:10]: Is this file needed? Can we replace it with logic in run.py

File: src/common/python/vms_layer/utils/date_helper.py [hunk 20]
```diff
@@ -0,0 +1,29 @@
...
+
+def epoch_To_date(epoch):
+    """Converts an epoch time to a date string."""
+    return time.strftime("%d-%m-%Y", time.localtime(epoch))
+
+
+def extract_quarters_from_date_range(start_date, end_date):
+    """Extracts the quarters from a date range maybe from different years."""
+    start_quater = int(start_date.split("-")[1]) // 3 + 1
+    print(start_quater)
```
Comment (vinsharma-wg) [inline src/common/python/vms_layer/utils/date_helper.py:18]: print statements should be removed

File: src/common/python/vms_layer/utils/date_helper.py [hunk 21]
```diff
@@ -0,0 +1,29 @@
...
+    years_in_between = int(end_date.split("-")[0]) - int(start_date.split("-")[0])
+    print(years_in_between)
```
Comment (vinsharma-wg) [inline src/common/python/vms_layer/utils/date_helper.py:20]: print statements should be removed. Replace it with logs

File: src/common/python/vms_layer/utils/rbac.py [hunk 22]
```diff
@@ -0,0 +1,22 @@
...
+            allowed_methods = rbac_data.get("ACCESS_CONTROL_LIST").get(group).get(resource)
+            if allowed_methods and http_method in allowed_methods:
+                return func(event, context)
+
+        return {
+            'statusCode': 403,
+            'body': 'Access Denied'
+        }
+    return wrapper
+
```
Comment (vinsharma-wg) [inline src/common/python/vms_layer/utils/rbac.py:22]: You can put this file in the validator folder

File: src/common/python/vms_layer/utils/validate_schema.py [hunk 23]
```diff
@@ -0,0 +1,26 @@
...
+                            "error": {"code": 400, "message": e.message},
+                            "status": "failure",
+                        }
+                    ),
+                }
+            return func(event, context)
+
+        return wrapper
+
+    return decorator
```
Comment (vinsharma-wg) [inline src/common/python/vms_layer/utils/validate_schema.py:26]: you can put this in validator folder

File: src/functions/login/app.py [hunk 24]
```diff
@@ -0,0 +1,28 @@
+import json
+import boto3
+import os
+
+
+def lambda_handler(event, context):
+    client = boto3.client("cognito-idp")
```
Comment (vinsharma-wg) [inline src/functions/login/app.py:7]: client should be outside the function or class scope. It will help in warm start up

File: src/functions/post_card/app.py [hunk 25]
```diff
@@ -0,0 +1,17 @@
...
+import os
+from vms_layer.utils.validate_schema import validate_schema
+from vms_layer.schemas.card_schema import card_schema
+from vms_layer.db_helper import DBHelper
+
+
+@validate_schema(schema=card_schema)
+def lambda_handler(event, context):
+    request_body = json.loads(event["body"])
+    db_helper = DBHelper(os.environ["DynamoDBTableName"])
```
Comment (vinsharma-wg) [inline src/functions/post_card/app.py:11]: environment variable should be in upper case like constants should be and they should be outside the function, just after the imports

File: template.yaml [hunk 26]
```diff
@@ -0,0 +1,485 @@
...
+    AWS SAM Template for making API Gateway with APIGatewayRole for invoking lambda functions.
+    Generate lambda functions according to docs\Visitor Management System Swagger 1.yaml
+
+Globals:
+    Function:
+        Timeout: 3
+        Runtime: python3.11
+        Environment:
+            Variables:
+                Stage: !Ref Stage
```
Comment (vinsharma-wg) [inline template.yaml:13]: Instead of stage use environment.

File: template.yaml [hunk 27]
```diff
@@ -0,0 +1,485 @@
...
+                                    !GetAtt VMSApprovalGET.Arn,
+                                    !GetAtt VMSApprovalPOST.Arn,
+                                    !GetAtt VMSApprovalIdGET.Arn,
+                                    !GetAtt VMSApprovalIdPATCH.Arn,
+                                ]
+
+    VMSApiGateway:
+        Type: AWS::Serverless::Api
+        Properties:
+            StageName: !Ref Stage
```
Comment (vinsharma-wg) [inline template.yaml:73]: Replace the Stage with environment.

File: docs/VMS_swagger.yaml [hunk 28]
```diff
@@ -0,0 +1,1727 @@
...
+                                - username
+                                - password
+                            type: object
+                            properties:
+                                username:
+                                    type: string
+                                password:
+                                    type: string
+            responses:
+                "200":
```
Comment (vinsharma-wg) [inline docs/VMS_swagger.yaml:42]: Please define it and refer it from components section

File: docs/VMS_swagger.yaml [hunk 29]
```diff
@@ -0,0 +1,1727 @@
...
+      description: Operations related to reports
+
+paths:
+    /login:
+        post:
+            summary: Login user
+            operationId: loginUser
+            tags:
+                - Authentication
+            requestBody:
```
Comment (vinsharma-wg) [inline docs/VMS_swagger.yaml:26]: Please define it and refer it from components section

File: docs/VMS_swagger.yaml [hunk 30]
```diff
@@ -0,0 +1,1727 @@
...
+        post:
+            summary: Forgot password
+            operationId: forgotPassword
+            tags:
+                - Authentication
+            requestBody:
+                description: User credentials for login
+                required: true
+                content:
+                    application/json:
```
Comment (vinsharma-wg) [inline docs/VMS_swagger.yaml:107]: Please define it and refer it from components section

File: docs/VMS_swagger.yaml [hunk 31]
```diff
@@ -0,0 +1,1727 @@
...
+                            required:
+                                - username
+                            type: object
+                            properties:
+                                username:
+                                    type: string
+                                new-password:
+                                    type: string
+            responses:
+                "200":
```
Comment (vinsharma-wg) [inline docs/VMS_swagger.yaml:118]: Please define it and refer it from components section

File: docs/VMS_swagger.yaml [hunk 32]
```diff
@@ -0,0 +1,1727 @@
...
+                                    type: string
+                                vistorPhotoBlob:
+                                    type: string
+                                idPhotoBlob:
+                                    type: string
+                                idProofNumber:
+                                    type: string
+            responses:
+                "201":
+                    description: Visitor created successfully
```
Comment (vinsharma-wg) [inline docs/VMS_swagger.yaml:303]: Please define it and refer it from components section

File: docs/VMS_swagger.yaml [hunk 33]
```diff
@@ -0,0 +1,1727 @@
...
+                - CognitoUserPool: []
+
+    /visitor/{id}:
+        get:
+            summary: Get visitor detail
+            operationId: getVisitorDetail
+            tags:
+                - Visitor
+            parameters:
+                - name: id
```
Comment (vinsharma-wg) [inline docs/VMS_swagger.yaml:361]: Please define it and refer it from components section

File: docs/VMS_swagger.yaml [hunk 34]
```diff
@@ -0,0 +1,1727 @@
...
+                  required: true
+                  schema:
+                      type: string
+            requestBody:
+                description: Updated visitor details
+                required: true
+                content:
+                    application/json:
+                        schema:
+                            type: object
```
Comment (vinsharma-wg) [inline docs/VMS_swagger.yaml:441]: Please define it and refer it from components section

File: docs/VMS_swagger.yaml [hunk 35]
```diff
@@ -0,0 +1,1727 @@
...
+                  schema:
+                      type: string
+                - name: approver
+                  in: query
+                  required: false
+                  schema:
+                      type: string
+
+            responses:
+                "200":
```
Comment (vinsharma-wg) [inline docs/VMS_swagger.yaml:559]: Please define it and refer it from components section

File: docs/VMS_swagger.yaml [hunk 36]
```diff
@@ -0,0 +1,1727 @@
...
+        post:
+            summary: Create new visit of Visitor
+            operationId: createVisit
+            tags:
+                - Visit
+            requestBody:
+                description: Visit details for creation
+                required: true
+                content:
+                    application/json:
```
Comment (vinsharma-wg) [inline docs/VMS_swagger.yaml:620]: Please define it and refer it from components section

File: cfn/iam_policy.yaml (hunk 2 above)
Comment (ahada-wg) [reply cfn/iam_policy.yaml:11]:
```
        MaxLength: 80
        Default: "lambda-iam-role"
    Environment:
```

Updated in PR#2

File: cfn/dynamodb.yaml (hunk 3 above)
Comment (ahada-wg) [reply cfn/dynamodb.yaml:8]:
```
        MaxLength: 80
        Default: "lambda-iam-role"
    Environment:
```

Updated in PR#2

File: cfn/cognito.yaml (hunk 5 above)
Comment (ahada-wg) [reply cfn/cognito.yaml:5]:
```
Parameters:
    Environment:
        Type: String
        Description: Environment of the deployment
        Default: dev
```

Updated in PR#2

File: cfn/dynamodb.yaml (hunk 6 above)
Comment (ahada-wg) [reply cfn/dynamodb.yaml:16]:
```
        Description: The name of the DynamoDB table
    Environment:
        Type: String
        Description: The name of the Environment
        AllowedValues:
            - dev
            - prod
```

Updated in PR#2

File: run.py (hunk 10 above)
Comment (ahada-wg) [reply run.py:119]: Change Done, Now the deployment will be successful at any point i.e. first deploy or any update.

File: cfn/static_content_bucket.yaml (hunk 9 above)
Comment (ahada-wg) [reply cfn/static_content_bucket.yaml:44]: Done!
Now as suggested, a new function will first check and extract outputs which are then passed as params to the required cfs.
```
        f"--parameter-overrides Environment={env_variables.get('ENVIRONMENT')} "
        f"BucketName={outputs.get('BucketName')} "
        f"DynamoDBTable={outputs.get('DynamoDBTableName')} "
        f"UserPoolId={outputs.get('UserPoolId')} "
        f"UserPoolClientId={outputs.get('UserPoolClientId')} "
        f"LambdaIAMRoleArn={outputs.get('LambdaIAMRoleArn')} "
    )
```

File: src/common/python/vms_layer/utils/date_helper.py (hunk 12 above)
Comment (umani-wg) [reply src/common/python/vms_layer/utils/date_helper.py:10]: Updated
```

def current_time_epoch():
    """Returns the current time in epoch format."""
    return int(time.time())


def date_to_epoch(date):
    """Converts a datetime string to epoch time."""
    return int(time.mktime(time.strptime(date, "%Y-%m-%dT%H:%M:%S")))


def epoch_to_date(epoch):
```
    """Converts an epoch time to a date string."""
    return time.strftime("%Y-%m-%dT%H:%M:%S", time.localtime(epoch))

File: src/common/python/vms_layer/utils/date_helper.py (hunk 13 above)
Comment (umani-wg) [reply src/common/python/vms_layer/utils/date_helper.py:5]: Done

File: cfn/iam_policy.yaml (hunk 14 above)
Comment (umani-wg) [reply cfn/iam_policy.yaml:47]: Updated and added in PR#2

File: run.py (hunk 15 above)
Comment (umani-wg) [reply run.py:20]: Sure, updated.

File: run.py (hunk 16 above)
Comment (umani-wg) [reply run.py:33]: Sure, updated.

File: run.py (hunk 17 above)
Comment (umani-wg) [reply run.py:42]: Sure, updated.

File: cfn/iam_policy.yaml (hunk 14 above)
Comment (ahada-wg) [reply cfn/iam_policy.yaml:47]: Permission for dynamodb added in iam role in PR#2,
Although in PR#1 we had a DynamoDbCRUD policy was directly applied to lambda in sam.

File: run.py (hunk 18 above)
Comment (umani-wg) [reply run.py:51]: Sure, updated.

File: samconfig.toml (hunk 19 above)
Comment (umani-wg) [reply samconfig.toml:10]: Replaced and updated in PR#2

File: src/common/python/vms_layer/utils/date_helper.py (hunk 20 above)
Comment (umani-wg) [reply src/common/python/vms_layer/utils/date_helper.py:18]: Removed all print statements.

File: src/common/python/vms_layer/utils/date_helper.py (hunk 21 above)
Comment (umani-wg) [reply src/common/python/vms_layer/utils/date_helper.py:20]: Removed all print statements.
